receiver/hostmetricsreceiver/internal/scraper/cpuscraper/internal/cpu_utilization_calculator.go

```diff
@@ -0,0 +1,102 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"fmt"
+
+	"github.com/shirou/gopsutil/v3/cpu"
+	"go.opentelemetry.io/collector/model/pdata"
+)
+
+// CPUUtilization stores the utilization percents [0-1] for the different cpu states
+type CPUUtilization struct {
+	CPU     string
+	User    float64
+	System  float64
+	Idle    float64
+	Nice    float64
+	Iowait  float64
+	Irq     float64
+	Softirq float64
+	Steal   float64
+}
+
+// CPUUtilizationCalculator calculates the cpu utilization percents for the different cpu states
+// It requires 2 []cpu.TimesStat and spend time to be able to calculate the difference
+type CPUUtilizationCalculator struct {
+	previousTime     pdata.Timestamp
+	previousCPUTimes []cpu.TimesStat
+}
+
+// Calculate calculates the cpu utilization for the different cpu states comparing previously
+// stored []cpu.TimesStat and time.Time and current []cpu.TimesStat and current time.Time
+// If no previous data is stored it will return empty slice of CPUUtilization and no error
+func (c *CPUUtilizationCalculator) Calculate(now pdata.Timestamp, cpuTimes []cpu.TimesStat) ([]CPUUtilization, error) {
+	var utilizationByCPU []CPUUtilization
+	elapsedSeconds := now.AsTime().Sub(c.previousTime.AsTime()).Seconds()
+
+	if c.previousCPUTimes != nil {
+		for _, previousCPUTime := range c.previousCPUTimes {
+			currentCPUTime, err := cpuTimeByCPU(previousCPUTime.CPU, cpuTimes)
+			if err != nil {
+				return []CPUUtilization{}, err
```

Comment:
Could we return `nil` and wrap the error?
```suggestion
				return nil, fmt.Errorf("getting time for cpu %d: %w", previousCPUTime.CPU, err)
```